src/runtime/borrowed_account.zig

```diff
@@ -39,85 +48,132 @@ pub const BorrowedAccount = struct {
-        const resize_delta: i64 = @intCast(length -| old_length);
-        const new_accounts_resize_delta = etc.accounts_resize_delta +| resize_delta;
+        const length_signed: i64 = @intCast(length);
+        const old_length_signed: i64 = @intCast(old_length);
+        const new_accounts_resize_delta = resize_delta +| length_signed -| old_length_signed;
 
         if (new_accounts_resize_delta > MAX_PERMITTED_ACCOUNTS_DATA_ALLOCATIONS_PER_TRANSACTION)
             return InstructionError.MaxAccountsDataAllocationsExceeded;
+
+        return null;
+    }
+
+    /// [agave] https://github.com/anza-xyz/agave/blob/c5ed1663a1218e9e088e30c81677bc88059cc62b/sdk/transaction-context/src/lib.rs#L825
+    pub fn setLamports(self: *BorrowedAccount, lamports: u64) InstructionError!void {
+        if (lamports < self.account.lamports and
+            !self.account.owner.equals(&self.context.program_id))
+        {
+            return InstructionError.ExternalAccountLamportSpend;
+        }
+
+        if (!self.context.is_writable)
+            return InstructionError.ReadonlyLamportChange;
+
+        if (self.account.executable)
+            return InstructionError.ExecutableLamportChange;
+
+        self.account.lamports = lamports;
+    }
+
+    /// [agave] https://github.com/anza-xyz/agave/blob/faea52f338df8521864ab7ce97b120b2abb5ce13/sdk/src/transaction_context.rs#L800
+    pub fn addLamports(self: *BorrowedAccount, lamports: u64) InstructionError!void {
+        self.account.lamports = std.math.add(
+            u64,
+            self.account.lamports,
+            lamports,
+        ) catch return InstructionError.ArithmeticOverflow;
+    }
+
+    /// [agave] https://github.com/anza-xyz/agave/blob/faea52f338df8521864ab7ce97b120b2abb5ce13/sdk/src/transaction_context.rs#L808
+    pub fn subtractLamports(self: *BorrowedAccount, lamports: u64) InstructionError!void {
+        self.account.lamports = std.math.sub(
+            u64,
+            self.account.lamports,
+            lamports,
+        ) catch return InstructionError.ArithmeticOverflow;
     }
 
-    /// Deserialize the account data into a type `T`.
+    /// [agave] https://github.com/anza-xyz/agave/blob/134be7c14066ea00c9791187d6bbc4795dd92f0e/sdk/src/transaction_context.rs#L817
+    pub fn constAccountData(self: BorrowedAccount) []const u8 {
+        return self.account.data;
+    }
+
+    /// [agave] https://github.com/anza-xyz/agave/blob/134be7c14066ea00c9791187d6bbc4795dd92f0e/sdk/src/transaction_context.rs#L823
+    pub fn mutableAccountData(self: BorrowedAccount) InstructionError![]u8 {
+        if (self.checkDataIsMutable()) |err| return err;
+        return self.account.data;
+    }
+
+    /// [agave] https://github.com/anza-xyz/agave/blob/faea52f338df8521864ab7ce97b120b2abb5ce13/sdk/src/transaction_context.rs#L885
+    pub fn setDataLength(
+        self: *BorrowedAccount,
+        allocator: std.mem.Allocator,
+        resize_delta: *i64,
+        new_length: usize,
```

Comment:
I'm a bit confused what resize_delta is for, seems agave just uses self.data.len inside the function

Reply:
Nevermind, I see it now :ok_hand: